File: src/jvmMain/java/resaver/pex/VDataID.java
package resaver.pex;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Objects;
import java.util.Set;
import java.util.function.Predicate;
import java.util.regex.Pattern;

import static resaver.pex.DataType.IDENTIFIER;

/**
 * VData that stores an identifier.
 */
public class VDataID extends VData {

    VDataID(TString val) {
        this.value = Objects.requireNonNull(val);
    }

    @Override
    public void write(ByteBuffer output) throws IOException {
        output.put((byte) this.getType().ordinal());
        this.value.write(output);
    }

    @Override
    public int calculateSize() {
        return 3;
    }

    @Override
    public void collectStrings(Set<TString> strings) {
        strings.add(this.value);
    }

    @Override
    public DataType getType() {
        return IDENTIFIER;
    }

    @Override
    public String toString() {
        //return String.format("ID[%s]", this.VALUE);
        return this.value.toString();
    }

    public TString getValue() {
        return this.value;
    }

    void setValue(TString val) {
        this.value = Objects.requireNonNull(val);
    }

    public boolean isTemp() {
        return TEMP_PATTERN.test(this.value.toString())
                && !AUTOVAR_PATTERN.test(this.value.toString())
                && !NONE_PATTERN.test(this.value.toString());
    }

    public boolean isAutovar() {
        return AUTOVAR_PATTERN.test(this.value.toString());
    }

    public boolean isNonevar() {
        return NONE_PATTERN.test(this.value.toString());
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 83 * hash + this.value.hashCode();
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null) {
            return false;
        } else if (getClass() != obj.getClass()) {
            return false;
        }
        final VDataID other = (VDataID) obj;
        return Objects.equals(this.value, other.value);
    }

    private TString value;
    static final Predicate<String> TEMP_PATTERN = Pattern.compile("^::.+$", Pattern.CASE_INSENSITIVE).asPredicate();
    static final Predicate<String> NONE_PATTERN = Pattern.compile("^::NoneVar$", Pattern.CASE_INSENSITIVE).asPredicate();
    static final Predicate<String> AUTOVAR_PATTERN = Pattern.compile("^::(.+)_var$", Pattern.CASE_INSENSITIVE).asPredicate();
}

File: src/jvmMain/kotlin/resaver/pex/VDataID.kt
package resaver.pex

import PlatformByteBuffer
import java.io.IOException
import java.util.function.Predicate
import java.util.regex.Pattern

/**
 * VData that stores an identifier.
 */
class VDataID internal constructor(`val`: TString?) : VData() {
    @Throws(IOException::class)
    override fun write(output: PlatformByteBuffer?) {
        output?.put(type.ordinal.toByte())
        value.write(output)
    }

    override fun calculateSize(): Int {
        return 3
    }

    override fun collectStrings(strings: MutableSet<TString?>?) {
        strings?.add(value)
    }

    override val type: DataType
        get() = DataType.IDENTIFIER

    override fun toString(): String {
        //return String.format("ID[%s]", this.VALUE);
        return value.toString()
    }

    val isTemp: Boolean
        get() = (TEMP_PATTERN.test(value.toString())
                && !AUTOVAR_PATTERN.test(value.toString())
                && !NONE_PATTERN.test(value.toString()))
    val isAutovar: Boolean
        get() = AUTOVAR_PATTERN.test(value.toString())
    val isNonevar: Boolean
        get() = NONE_PATTERN.test(value.toString())

    override fun hashCode(): Int {
        var hash = 7
        hash = 83 * hash + value.hashCode()
        return hash
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        } else if (other == null) {
            return false
        } else if (javaClass != other.javaClass) {
            return false
        }
        val other2 = other as VDataID
        return value == other2.value
    }

    var value: TString = `val`!!

    companion object {
        val TEMP_PATTERN: Predicate<String> = Pattern.compile("^::.+$", Pattern.CASE_INSENSITIVE).asPredicate()
        val NONE_PATTERN: Predicate<String> = Pattern.compile("^::NoneVar$", Pattern.CASE_INSENSITIVE).asPredicate()
        val AUTOVAR_PATTERN: Predicate<String> = Pattern.compile("^::(.+)_var$", Pattern.CASE_INSENSITIVE).asPredicate()
    }

}